File: src/main/java/com/ptaf/ai/index/StepDefinitionIndex.java
package com.ptaf.ai.index;

import com.ptaf.ai.StepPatternExtractor;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Index of known step definition patterns extracted from Java annotation values.
 */
public final class StepDefinitionIndex {
    private final List<String> knownSteps;

    public StepDefinitionIndex(List<String> knownSteps) {
        this.knownSteps = List.copyOf(knownSteps);
    }

    public List<String> knownSteps() {
        return knownSteps;
    }

    public static StepDefinitionIndex build(Path projectRoot, List<String> stepDefinitionPaths) throws IOException {
        Set<String> ordered = new LinkedHashSet<>();
        for (String relativePath : stepDefinitionPaths) {
            if (relativePath == null || relativePath.isBlank()) {
                continue;
            }
            Path root = projectRoot.resolve(relativePath).normalize();
            if (!Files.isDirectory(root)) {
                continue;
            }
            for (Path javaFile : listJavaFiles(root)) {
                String source = Files.readString(javaFile, StandardCharsets.UTF_8);
                ordered.addAll(StepPatternExtractor.fromJavaSource(source));
            }
        }
        return new StepDefinitionIndex(new ArrayList<>(ordered));
    }

    private static List<Path> listJavaFiles(Path root) throws IOException {
        try (Stream<Path> walk = Files.walk(root)) {
            return walk
                    .filter(Files::isRegularFile)
                    .filter(path -> path.toString().toLowerCase().endsWith(".java"))
                    .sorted(Comparator.comparing(Path::toString))
                    .collect(Collectors.toList());
        }
    }
}

File: python/ptaf/ai/index/step_definition_index.py
"""Index of known step definition patterns extracted from Python step files."""

from __future__ import annotations

from pathlib import Path

from ptaf.ai.step_pattern_extractor import from_python_source


class StepDefinitionIndex:
    def __init__(self, known_steps: list[str]) -> None:
        self._known_steps = list(known_steps)

    def known_steps(self) -> list[str]:
        return self._known_steps

    @classmethod
    def build(cls, project_root: Path, step_definition_paths: list[str]) -> StepDefinitionIndex:
        ordered: list[str] = []
        seen: set[str] = set()
        for relative_path in step_definition_paths:
            if not relative_path or not relative_path.strip():
                continue
            root = (project_root / relative_path).resolve()
            if not root.is_dir():
                continue
            for py_file in sorted(root.rglob("*.py")):
                if not py_file.is_file():
                    continue
                source = py_file.read_text(encoding="utf-8")
                for pattern in from_python_source(source):
                    if pattern not in seen:
                        seen.add(pattern)
                        ordered.append(pattern)
        return cls(ordered)
